Add compute lecture (#5)
open-education-hub/operating-systems: content/chapters/compute/lecture/slides.md

20733 pts/1    00:00:18 node
20796 pts/1    00:00:25 node
20915 ?        00:00:00 init
20916 ?        00:00:00 init
20917 pts/4    00:00:00 bash
20964 pts/4    00:00:00 python3
20981 ?        00:00:00 init
20982 ?        00:00:00 init
20983 pts/5    00:00:00 bash
21045 pts/5    00:00:00 ps
```

Way more processes than cores!

---

## 1. Processes

* A process is a **running program**
    * It is a collection of resources/abstractions (Memory, CPU, I/O)
* An application can spawn multiple processes
* The OS allows each thread to run on a **core** for a few milliseconds (time slice)
    * See [section 4](#2-threads---backstory) for details regarding threads
* Then the OS pauses the thread and replaces it with another one (context switch)

----

### Scheduling - Overview

![Scheduling Overview](media/scheduling-overview.svg)

----

### What We Want

* Run more apps than cores
* Efficient scheduling
* Encapsulation and separation of tasks:
    * VAS: isolate data
    * CPU: computations, threads
    * I/O: network packets, files

----

### What We Don't Want

* Unresponsive processes
* Processes monopolizing resources
* System freezes

---

### Process Address Space

> * Each process has its own VAS:

Reviewer: ```suggestion
* Each process has its own VAS (*Virtual Address Space*):
```

    * Isolation
    * The process "thinks" it uses the entire address space

----

> ### Process Address Space

Reviewer: ```suggestion
### Process Virtual Address Space
```

> ![Process VAS](media/vas.svg)

Reviewer: Remove GOT and PLT from the diagram. They are not really dynamic and they are too much at this point.

---

> ## Process Attributes - from User Space

Reviewer: Why "from User Space" - I think this is a bit confusing.

Author: I meant to describe the attributes from 2 perspectives: what the user space can see and what the kernel can see: the PCB

* PID: numeric identifier
* PPID: parent's PID (one parent per process)
* Running time

> * State: we will discuss this later

Reviewer: ```suggestion
* threads
* memory map
* open files
```

20658 20653   604   2612    05:43:41 S+                \_ sh
20661 20658 74072 931768    05:43:41 Sl+                   \_ node
20681 20661 44316 614424    05:43:40 Sl+                       \_ node
20722 20661 39804 828152    05:43:40 Sl+                       \_ node
20733 20661 147004 845832   05:43:40 Sl+                       \_ node
20796 20733 119336 959952   05:43:38 Sl+                           \_ node
21149 20733 41080 581504    04:31:20 Sl+                           \_ node
20672     1   176    988    05:43:40 Ss   init
20673 20672   176    988    05:43:40 S     \_ init
20674 20673 41556 585172    05:43:40 Ssl+      \_ node
20701     1   176    988    05:43:40 Ss   init
20702 20701   176    988    05:43:40 S     \_ init
20703 20702 48060 590308    05:43:40 Ssl+      \_ node
20915     1   176    988    05:24:46 Ss   init
20916 20915   176    988    05:24:46 S     \_ init
20917 20916  5268  10188    05:24:46 Ss        \_ bash
20964 20917 19840 516296    05:24:32 S+            \_ python3
20981     1   176    988    05:22:54 Ss   init
20982 20981   176    988    05:22:54 S     \_ init
20983 20982  5100  10056    05:22:54 Ss+       \_ bash
```

----

> ### Process Attributes - from Kernel Space

Reviewer: ```suggestion
### Process Attributes (internals)
```

* PCB (Process Control Block):

>     * Data structure in the kernel - `task_struct`

Reviewer: ```suggestion
    * Data structure in the kernel - `task_struct` (in Linux)
```

>     * All information about the process (from `ps`, `/proc`):

Reviewer: Why is there a colon (`:`)?

Author: It was meant to be a hint to look at the next slide, but in reality, it looks stupid. I've removed it.

----

### Process Attributes - from Kernel Space

The most important fields in `task_struct`:
```c [2 | 4 | 6 | 8 - 9 | 11 - 13 | 15 - 16 | 18 - 19 | 21 | 23 | 25 - 26 | 28 - 29]
struct task_struct {
        unsigned int                    __state;

        void                           *stack;

        unsigned int                    flags;

        int                             on_cpu;
        int                             prio;

        /* Scheduler information */
        struct sched_entity             se;
        const struct sched_class        *sched_class;

        /* The VAS: memory mappings */
        struct mm_struct                *mm;

        int                             exit_state;
        int                             exit_code;

        pid_t                           pid;

        struct task_struct __rcu        *parent;

        /* Child processes */
        struct list_head                children;

        /* Open file information */
        struct files_struct             *files;
};
```

---

> ### `fork`

Reviewer: I think this is too sudden. I would add a section called "Process Lifetime". I would state the main phases of a process and user actions: creation / birth, runtime (and live investigation by the user via ps, top etc.), termination. And the actions done by the parent process corresponding to this.

I can easily see a diagram for this.

You can then do a slide for each state with a short summary. Then the next section would be creation / birth / fork / spawn.

Reviewer: Even more so, I think fork is too much here, it's more of an internals aspect. I would first go discussing more on how you start processes, using an executable, understanding the interface.

See the demos:
```
razvan@yggdrasil:~/school/so/curs.git$ git remote show origin
* remote origin
  Fetch URL: so@elf.cs.pub.ro:git-repos/curs.git
  Push  URL: so@elf.cs.pub.ro:git-repos/curs.git
  HEAD branch: master
  Remote branch:
    master tracked
  Local branch configured for 'git pull':
    master merges with remote master
  Local ref configured for 'git push':
    master pushes to master (up to date)
razvan@yggdrasil:~/school/so/curs.git$ ls
curs-01-demo  curs-04-demo  curs-07-demo      curs-09-demo      curs-11-demo  exec-process-demo      make-clean    publish-demos
curs-02-demo  curs-05-demo  curs-07-demo.zip  curs-09-demo.zip  curs-12-demo  exec-process-demo.zip  mem-sec.wiki  sync.wiki
curs-03-demo  curs-06-demo  curs-08-demo      curs-10-demo      curs-13-demo  exec-process.wiki      note-curs
razvan@yggdrasil:~/school/so/curs.git$ ls curs-04-demo/
nice  pipe  pipe-exec  utils  yield
razvan@yggdrasil:~/school/so/curs.git$ ls curs-03-demo/
ctxt-switch  executor  fork-file-pointer  orphan-zombie  utils
razvan@yggdrasil:~/school/so/curs.git$ ls curs-03-demo/executor/
executor       executor_ansi.c  executor.c  executor.py     executor_spawn.c  executor_system    Makefile  printer.d
executor_ansi  executor_ansi.o  executor.o  executor_spawn  executor_spawn.o  executor_system.c  printer   printer.o
```

Also check the recording of my process lecture for a narrative: https://ctipub.sharepoint.com/:f:/s/03-ACS-L-A3-S2-SO-CA-CB-CC-2021/Epei_BOAME9PkX5jLuCl1owBj3upEwzHxDHQU-014wRHqA?e=Bugjvv

* Spawns child process
* Child process starts execution right after the call to `fork`
* Returns twice
    * `0` in child process
    * Child's PID in parent process
[Quiz](./quiz/process-creation.md)
* Initially child and parent share the same PAS (Physical Address Space)
* Child's writable pages are marked **copy-on-write**

----

### Copy-on-write

> When `fork` returns, the child inherits the parent's page table.

Reviewer: ```suggestion
When `fork` returns, the child duplicates the contents of the parent's page table.
```

![Copy-on-write after fork()](./media/copy-on-write-initial.svg)

----

### Copy-on-write

> * Writable pages are **copied and modified** when written: `demo/copy-on-write/copy_on_write_overhead.c`.

Reviewer: ```suggestion
* Writable pages are **copied and modified** when written: `demo/copy-on-write/copy_on_write_overhead.c`
```

* [Quiz](./quiz/sections-always-shared.md)

![Copy-on-write after Write](./media/copy-on-write-final.svg)

----

> ### Copy-on-write Overheard

Reviewer: This doesn't render properly.

Author: It seems the `$` character was somehow interpreted either as HTML or as a genuine command such as `curl` whose output was embedded in the rendering :)))

Moral of the story: don't use backticks instead of `$(...)` around commands.

    * Child's PID in parent process
[Quiz](./quiz/process-creation.md)
* Initially child and parent share the same PAS (Physical Address Space)
* Child's writable pages are marked **copy-on-write**

----

### Copy-on-write

When `fork` returns, the child inherits the parent's page table.

![Copy-on-write after fork()](./media/copy-on-write-initial.svg)

----

### Copy-on-write

* Writable pages are **copied and modified** when written: `demo/copy-on-write/copy_on_write_overhead.c`.
* [Quiz](./quiz/sections-always-shared.md)

![Copy-on-write after Write](./media/copy-on-write-final.svg)

----

### Copy-on-write Overheard

* The overhead from writing data comes mostly from minor page faults
```
student@os:~/.../compute/lecture/demo/copy-on-write$ ./copy_on_write_overhead 
 * Child prcess started
 -- Press ENTER to continue ...
Time for reading 131072 pages: 30 ms
 * Child process read pages
 -- Press ENTER to continue ...
Time for writing to 131072 pages: 222 ms
[...]
```

* The child process performs exactly 131072 **minor** page faults when writing data
```
student@os:~$ cat /proc/$(pidof copy_on_write_overhead)/stat | cut -d ' ' -f 10  # before writing
22

> 

Author: This one is alright. We'll leave one empty line between subsequent commands to distinguish them better.

student@os:~$ cat /proc/$(pidof copy_on_write_overhead)/stat | cut -d ' ' -f 10  # after writing
131094
```

---

## 2. Threads - Backstory

* You have an application that applies filters to a set of images
* Processes need to share all images
    * Difficult because the images (like all data) are marked copy-on-write
* **Solution:** use threads
    * No copy-on-write: threads share the heap and data sections

----

### 2. Threads - Definitions

* Each process has at least one thread
* Lightweight processes (LWP)
* Scheduling unit

---

### Thread - Resources

![Thread Resources](media/process-vs-thread.png)

----

### App, Process, Thread - The Bigger Picture

* [Quiz](./quiz/threads-shared-data.md)

![Application Architecture](media/app-process-thread.svg)

----

### `fork` and Threads

* `fork` also creates a new thread...
* ... but part of a different process

![fork - new Thread](./media/fork-thread.svg)

---

### Process vs Thread

> 

Author: I think most of the implementation effort is somewhat common. If you only had processes and not separate threads (i.e. a process only contains a single thread), you would run into the same issues. The problem actually comes from parallelism itself rather than threads or processes per se.

```

---

### Context Switch - Overhead

* Causes of overhead:
    * running the scheduler itself
    * flushing the TLB of the evicted thread
* Faster context switches between threads of the same process because their TLB is not flushed

---

### Scheduling - Ideals

* **Interactivity:** decrease time spent in READY
    * low waiting time
    * also known as **fairness**
* **Throughput:** finish threads as quickly as possible
    * low turnaround time

----

### Scheduling - Reality

* There is no free lunch
* Tradeoff between interactivity and throughput

----

### Types of Schedulers

* **Cooperative:** only allow voluntary context switches
    * Advantage: Simple to implement
    * Disadvantage: starvation - a thread might never yield its CPU core
* **Preemptive:** also allows involuntary context switches
    * Each process can run at most a certain _time slice_ / _quantum_
    * When it expires, a context switch is triggered

----

### Scheduling Algorithms - The Bigger Picture

* **Round-Robin**:
    * Push each new thread to a queue
    * Every time slice, dequeue one thread, run it, then reenqueue it
    * Threads run in the order in which they're spawned

> * **Completely Fair Scheduler (CFS)**:

Author: I reworded it to make it fit. I hope it's still coherent. I wish all this info stays on the same slide to make an easier comparison between Round-Robin and CFS.

```

* Notice that threads reach the barrier and end in arbitrary order

---

### Condition Variables

* More complex synchronization logic
* Similar to but more flexible than semaphores
* A thread is suspended until a condition (logical predicate) is met
* Always used with mutexes (evaluating the condition is **not atomic**)

----

### Condition Variables - API

* `wait`: suspend the thread until the condition is met
    * need to acquire the lock before waiting
    * this is needed to check the condition in a thread-safe manner
    * `wait` releases the lock
```python
cond.acquire()
cond.wait()
```
* `notify`: wake up **one** thread waiting on the condition
* `notify_all`: wake up **all** threads waiting on the condition

----

### Condition Variables - Demo

```
student@os:~/.../compute/lecture/demo/condition$ python3 condition.py
Waiting thread: waiting for main thread...
Main thread: type "notify" to notify waiting thread

waiting thread is sleeping until woken up by main thread

notify
Waiting thread: notified by main thread
```

---

### Thread-safety and Reentrancy

* **Thread-safe function:** can be called simultaneously from multiple threads without race conditions

> * **Reentrant function:** can be preempted at any time and called from the same thread

Author: I think this one is pretty obvious. Any preemption is useless if it results in the preempted thread behaving differently.

clone(child_stack=NULL, flags=CLONE_CHILD_CLEARTID|CLONE_CHILD_SETTID|SIGCHLD, child_tidptr=0x7f7e83aa4810) = 5302
student@os:~/.../compute/lecture/demo/create-thread$ strace -e clone ./create_thread
clone(child_stack=0x7f9ea7df0fb0, flags=CLONE_VM|CLONE_FS|CLONE_FILES|CLONE_SIGHAND|CLONE_THREAD|CLONE_SYSVSEM|CLONE_SETTLS|CLONE_PARENT_SETTID|CLONE_CHILD_CLEARTID, tls=0x7f9ea7df1700, child_tidptr=0x7f9ea7df19d0) = 5389
```
* From the flags, the parent and the child **thread** share:
    * `CLONE_VM`: VAS (no copy-on-write)
    * `CLONE_FS`: FS information (root, umask, cwd)
    * `CLONE_FILES`: File descriptor table
    * `CLONE_SIGHAND`: Signal handlers

---

### The Suspended States

![Extended Thread Diagram](./media/thread-states-extended.svg)

----

### The Suspended States

* **SUSPENDED READY / WAITING:** placed on secondary storage (**swap area**) due to lack of RAM
    * Brought back to READY / WAITING when more RAM is available
* Threads can move from SUSPENDED WAITING to SUSPENDED READY when I/O completes
* Impossible to move from SUSPENDED READY to SUSPENDED WAITING (threads cannot perform I/O calls without running)

---

> ### User Level Threads

Reviewer 2: ```suggestion
### Kernel Level Threads
```
The content seems to be about KLT.
